### PR TITLE
Add JWT Token Authentication

### PR DESCRIPTION
**Pull request by jonathangreen**
This adds JWT token authentication to Islandora. We send these tokens
along with broadcast messages, and can use them to authenticate with
Fedora as well as call back into Islandora.

This is part of:
https://github.com/Islandora-CLAW/CLAW/issues/520

### CODE REVIEW

**Comment by jonathangreen** [comment]
Travis is having issues with the dependancy for the JWT module, since composer runs after we try to enable modules with drush. I'll try to untangle it. Some fires in other places right now though, so it'll have to wait to next week to get untangled.

**Comment by DiegoPino** [comment]
@jonathangreen someone has been very active lately!
https://github.com/gabesullice/jwt/pull/5 - https://github.com/gabesullice/jwt/pull/8! (also see 6 and 7)
My respects! :bowtie:

**Comment by DiegoPino** [comment]
@jonathangreen not sure if this helps: you could deploy your own TravisCI local environment to test this combinatory of changes until one passes.
Instructions: https://docs.travis-ci.com/user/common-build-problems/#Running-a-Container-Based-Docker-Image-Locally

Travis keeps docker(official by them) images https://quay.io/organization/travisci
In specific i think `travis-php`, ubuntu 12.04, one can be used out of the box or tuned 
```Shell
docker pull quay.io/travisci/travis-php
````
If not the  `travis-standard`
Ubuntu 14 ones start with `ci-`(i think... its saturday morning)

the issue is how you pass the build commands/run them. Basically you need to run them manually or make a script like the thing i did for the failing islandora newspaper stuff.

**Comment by DiegoPino** [comment]
@jonathangreen also: if you add the JWT dependency [here](https://github.com/Islandora-CLAW/drupal-project/blob/8.x-1.x/composer.json)? Does that change something?

**Comment by jonathangreen** [comment]
Thanks Diego. That is very helpful.

**Comment by jonathangreen** [comment]
Decided to break the travis stuff out into a different pull request. 

Once this is in: 
https://github.com/Islandora-CLAW/islandora/pull/39

This should test cleanly.

**Comment by whikloj** [comment]
@jonathangreen if you want to rebase this now that the changed travis setup is merged, we can see if it passes.

**Comment by jonathangreen** [comment]
Will do.

File: src/EventSubscriber/JwtEventSubscriber.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+use Drupal\Core\Entity\EntityTypeManagerInterface;
+use Drupal\Core\Session\AccountInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+/**
+ * Class JwtEventSubscriber.
+ *
+ * @package Drupal\islandora\EventSubscriber
+ */
+class JwtEventSubscriber implements EventSubscriberInterface {
```
Comment (dannylamb) [inline src/EventSubscriber/JwtEventSubscriber.php]: This guy probably deserves some tests.  I can explore and see how terrible the mocking will be.  Possible PR pending.

**Comment by whikloj** [comment]
This works for me, took a bit to work out and I had problems due to https://github.com/Islandora-CLAW/claw-jsonld/pull/8, but now I can see the items in the 3store using only JWT authentication.

👍

**Comment by whikloj** [comment]
@DiegoPino can I merge this, or do you still have concerns?

**Comment by whikloj** [comment]
@dannylamb sorry I forgot you were going to add tests. I can wait.

**Comment by DiegoPino** [comment]
@whikloj i approved, but Dani wanted to  add tests. So you are free to decide what order is best.

**Comment by dannylamb** [comment]
I'll add tests. It needs to become habitual that we add them upfront.

**Comment by dannylamb** [comment]
https://github.com/jonathangreen/islandora-claw/pull/1/files

**Comment by DiegoPino** [comment]
@dannylam why we are moving to a lower level class storage instead of the previous, higher level interface? :smile: i want to learn !

**Comment by jonathangreen** [comment]
Merged those changes.

**Comment by dannylamb** [comment]
@DiegoPino  In order to hit all of the code, I needed to mock the user storage to return dummy values in a couple of situations, such as this: https://github.com/jonathangreen/islandora-claw/pull/1/files#diff-277bcde9c6229a2b678538808655c0b2R106

Injecting the user storage directly makes it easier to do that.  Plus we are only using the EntityTypeManagerInterface to look up that storage.  If we were using other functionality, I would've kept it and attempted to double mock it.  E.g. prophecise an entityTypeManager that returns a prophecised user storage.

**Comment by dannylamb** [comment]
@jonathangreen Do we need to rebase this to pull in recent Travis changes?  He seems unhappy setting himself up.

**Comment by jonathangreen** [comment]
Weird yah. Let me try a quick rebase.

**Comment by DiegoPino** [comment]
@dannylamb ok cool Dany, thanks for the explanation (with prophecies you mean actually https://github.com/phpspec/prophecy)

I really don't want to go into Class prophetizing talk so Ok(who knows what will eventually happen in the future anyway, reminds me of that Interface versus Abstract class 😨 discussion !) Hopefully nobody installs a user provider swapping module that does not inherit from storage also, but that would be prophetizing ! 🙈  arghh

**Comment by jonathangreen** [comment]
I rebased but we seem to have angered the travis gods somehow. I'm in meetings for the next couple hours, but I can try to sort out whats up with travis after that.

**Comment by jonathangreen** [comment]
This needs to be rebased once #43 goes in so that travis runs smoothly.

**Comment by jonathangreen** [comment]
Rebased and pushed on top of new travis commits.

**Comment by jonathangreen** [comment]
@dannylamb looks like tests are running okay now, but maybe there is an issue with the test mocks as the test seems to be failing.

**Comment by dannylamb** [comment]
Hrm... `Call to undefined method: Drupal\jwt\Authentication\Event\JwtAuthGenerateEvent::isValid()`  That's weird.  I only call that functin on `JwtAuthValidateEvent`'s.  Let me see what's up.

**Comment by dannylamb** [comment]
Yarg, bad variable names from some last minute copy/pasta.  Let me fix that for you.

**Comment by dannylamb** [comment]
All good now.  Sorry about that.

**Comment by jonathangreen** [comment]
Updated the JWT structure to reflect the limitations of Java JWT libraries. I think this is ready to merge whenever anyone thinks its ready.

**Comment by jonathangreen** [comment]
Looks like the version of PHPCS updated with some new sniffs that is making this pull fail...

**Comment by dannylamb** [comment]
Ugh.  I can tackle this for you.

**Comment by dannylamb** [comment]
New phpcs is pretty smart.

```
----------------------------------------------------------------------
FOUND 0 ERRORS AND 1 WARNING AFFECTING 1 LINE
----------------------------------------------------------------------
 86 | WARNING | Possible useless method overriding detected
----------------------------------------------------------------------
```

<image src="http://vignette1.wikia.nocookie.net/jurassicpark/images/5/53/Raptor_-_Clever_Girl.gif/revision/latest?cb=20140408111011"/>

**Pull request merged**